internal/component/pyroscope/ebpf/ebpf_linux.go

```diff
@@ -104,48 +103,74 @@ type Component struct {
 	debugInfo     DebugInfo
 	debugInfoLock sync.Mutex
 	metrics       *metrics
+
+	healthMut sync.RWMutex
+	health    component.Health
 }
 
 func (c *Component) Run(ctx context.Context) error {
-	err := c.session.Start()
-	if err != nil {
-		return fmt.Errorf("ebpf profiling session start: %w", err)
-	}
-	defer c.session.Stop()
-
-	var g run.Group
-	g.Add(func() error {
-		collectInterval := c.args.CollectInterval
-		t := time.NewTicker(collectInterval)
-		defer t.Stop()
-		for {
-			select {
-			case <-ctx.Done():
-				return nil
-			case newArgs := <-c.argsUpdate:
-				c.args = newArgs
-				c.session.UpdateTargets(targetsOptionFromArgs(c.args))
-				c.metrics.targetsActive.Set(float64(len(c.targetFinder.DebugInfo())))
-				err := c.session.Update(convertSessionOptions(c.args, c.metrics))
-				if err != nil {
-					return nil
-				}
-				c.appendable.UpdateChildren(newArgs.ForwardTo)
-				if c.args.CollectInterval != collectInterval {
-					t.Reset(c.args.CollectInterval)
-					collectInterval = c.args.CollectInterval
-				}
-			case <-t.C:
-				err := c.collectProfiles()
+	started := false
+
+	collectInterval := c.args.CollectInterval
+	t := time.NewTicker(collectInterval)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case newArgs := <-c.argsUpdate:
+			// ensure there are no other updates queued. this might happen if the collection takes a very long time
+			newArgs = getLatestArgsFromChannel(c.argsUpdate, newArgs)
+
+			// update targets
+			c.args = newArgs
+			c.session.UpdateTargets(targetsOptionFromArgs(c.args))
+			c.metrics.targetsActive.Set(float64(len(c.targetFinder.DebugInfo())))
+			err := c.session.Update(convertSessionOptions(c.args, c.metrics))
+			if err != nil {
+				level.Error(c.options.Logger).Log("msg", "failed to update profiling session", "err", err)
+				c.reportUnhealthy(err)
+				continue
+			}
+			c.appendable.UpdateChildren(newArgs.ForwardTo)
+			if c.args.CollectInterval != collectInterval {
+				t.Reset(c.args.CollectInterval)
+				collectInterval = c.args.CollectInterval
+			}
+		case <-t.C:
+			if !started {
+				err := c.session.Start()
```

Comment:
I imagine main of those errors are permanent (e.g. not enough permissions, we probably should handle those by stopping the ticker. Wdyt @korniltsev

Reply:
Now implemented https://github.com/grafana/alloy/pull/3381/commits/19591e05769f961573d2888daae544a4bcd7c509